boot/seal_test.go

```diff
@@ -349,39 +420,42 @@ func (s *sealSuite) TestResealKeyToModeenv(c *C) {
 			runKernel := bootloader.NewBootFile(filepath.Join(rootdir, "var/lib/snapd/snaps/pc-kernel_500.snap"), "kernel.efi", bootloader.RoleRunMode)
 			runKernel2 := bootloader.NewBootFile(filepath.Join(rootdir, "var/lib/snapd/snaps/pc-kernel_600.snap"), "kernel.efi", bootloader.RoleRunMode)
 
-			c.Assert(params.ModelParams[0].EFILoadChains[2:4], DeepEquals, []*secboot.LoadChain{
-				secboot.NewLoadChain(shim,
-					secboot.NewLoadChain(grub,
-						secboot.NewLoadChain(runGrub,
-							secboot.NewLoadChain(runKernel)),
-						secboot.NewLoadChain(runGrub2,
-							secboot.NewLoadChain(runKernel)),
-					)),
-				secboot.NewLoadChain(shim2,
-					secboot.NewLoadChain(grub,
-						secboot.NewLoadChain(runGrub,
-							secboot.NewLoadChain(runKernel)),
-						secboot.NewLoadChain(runGrub2,
-							secboot.NewLoadChain(runKernel)),
-					)),
-			})
-
-			c.Assert(params.ModelParams[0].EFILoadChains[4:], DeepEquals, []*secboot.LoadChain{
-				secboot.NewLoadChain(shim,
-					secboot.NewLoadChain(grub,
-						secboot.NewLoadChain(runGrub,
-							secboot.NewLoadChain(runKernel2)),
-						secboot.NewLoadChain(runGrub2,
-							secboot.NewLoadChain(runKernel2)),
-					)),
-				secboot.NewLoadChain(shim2,
-					secboot.NewLoadChain(grub,
-						secboot.NewLoadChain(runGrub,
-							secboot.NewLoadChain(runKernel2)),
-						secboot.NewLoadChain(runGrub2,
-							secboot.NewLoadChain(runKernel2)),
-					)),
-			})
+			switch resealKeysCalls {
+			case 1:
+				c.Assert(params.ModelParams[0].EFILoadChains[2:4], DeepEquals, []*secboot.LoadChain{
+					secboot.NewLoadChain(shim,
+						secboot.NewLoadChain(grub,
+							secboot.NewLoadChain(runGrub,
+								secboot.NewLoadChain(runKernel)),
+							secboot.NewLoadChain(runGrub2,
+								secboot.NewLoadChain(runKernel)),
+						)),
+					secboot.NewLoadChain(shim2,
+						secboot.NewLoadChain(grub,
+							secboot.NewLoadChain(runGrub,
+								secboot.NewLoadChain(runKernel)),
+							secboot.NewLoadChain(runGrub2,
+								secboot.NewLoadChain(runKernel)),
+						)),
+				})
+
+				c.Assert(params.ModelParams[0].EFILoadChains[4:], DeepEquals, []*secboot.LoadChain{
+					secboot.NewLoadChain(shim,
+						secboot.NewLoadChain(grub,
+							secboot.NewLoadChain(runGrub,
+								secboot.NewLoadChain(runKernel2)),
+							secboot.NewLoadChain(runGrub2,
+								secboot.NewLoadChain(runKernel2)),
+						)),
+					secboot.NewLoadChain(shim2,
+						secboot.NewLoadChain(grub,
+							secboot.NewLoadChain(runGrub,
+								secboot.NewLoadChain(runKernel2)),
+							secboot.NewLoadChain(runGrub2,
+								secboot.NewLoadChain(runKernel2)),
+						)),
+				})
+			}
```

Comment:
this should have a default too I think

Reply:
When I add this it starts to fail it seems :(

Reply:
It's because the recovery boot chains are common for both keys and are checked in line 408 https://github.com/snapcore/snapd/pull/9577/files/74e5eaef07bc091e158d0c2b961fb406f4517847#diff-fda1ca0460d96a5f1f24e0fef2c6e29bdd73e9a28c4ddee02a01c14c3ced24d8R408

Now, to double check, the run mode data key is supposed to sealed to both recovery and run mode load chains, right?
While the fallback keys are sealed only to the recovery load chains?

Reply:
@bboozzoo 

> Now, to double check, the run mode data key is supposed to sealed to both recovery and run mode load chains, right?

yes because the first key we try to use to unseal / decrypt ubuntu-data in recover mode is the one on ubuntu-boot, if that fails then we fallback to the one on ubuntu-seed (which is the fallback key aiui)

> While the fallback keys are sealed only to the recovery load chains?

I think so, that's my understanding at least, is that we only intend to use the fallback keys from recovery (and at least that's what the uc20 save TODO's doc says)

Reply:
> yes because the first key we try to use to unseal / decrypt ubuntu-data in recover mode is the one on ubuntu-boot, if that fails then we fallback to the one on ubuntu-seed (which is the fallback key aiui)

just a note to myself - all keys exist on ubuntu-seed, we don't put any keys on ubuntu-boot currently